File: DSAlgorithms/Leetcode/c++/arrays_hash/304-range_sum_query_2d.cpp
class NumMatrix {
public:
    std::vector<std::vector<int>> prefix_sum;
    NumMatrix(vector<vector<int>>& matrix) {
       int rows = matrix.size(); 
       int cols = matrix[0].size();
       prefix_sum = matrix;

       //calc prefixsums
       for(int row = 0; row < rows; row++)
       {
           int prefix = 0;
           int above = 0;
           for(int col = 0; col < cols; col++)
           {
               prefix += matrix[row][col];
               if(row > 0)
                    above = prefix_sum[row - 1][col];
                prefix_sum[row][col] = prefix + above;
           }
       }
       for(int i = 0; i < rows; i++)
       {
           for(int j = 0; j < cols; j++)
                {
                    std::cout<<prefix_sum[i][j]<<" ";
                }
       std::cout<<'\n';
       }
    }
    
    int sumRegion(int row1, int col1, int row2, int col2) {
        int bottom_right = prefix_sum[row2][col2];
        int above = 0;
        int left = 0;
        int top_left = 0;
        if(row1 > 0)
             above = prefix_sum[row1 - 1][col2];
        if(col1 > 0)
            left = prefix_sum[row2][col1 - 1];
        if(row1 > 0 && col1 > 0)
            top_left = prefix_sum[row1 - 1][col1 - 1];

        return bottom_right - above - left + top_left;

    }
};

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix* obj = new NumMatrix(matrix);
 * int param_1 = obj->sumRegion(row1,col1,row2,col2);
 */

File: DSAlgorithms/Leetcode/ts/src/arrays_hash/304-range_sum_query-2d.ts
class NumMatrix {
    prefix_sum: number[][] = [];
    constructor(matrix: number[][]) {
        const rows: number = matrix.length;
        const cols: number = matrix[0].length;
        this.prefix_sum = matrix;
        for(let row = 0; row < rows; row++)
        {
            let prefix = 0;
            let above = 0;
            for(let col = 0; col < cols; col++)
            {
                prefix += matrix[row][col];
                if(row > 0)
                    above = this.prefix_sum[row -1][col];
                this.prefix_sum[row][col] = prefix + above;
            }
        }
    }

    sumRegion(row1: number, col1: number, row2: number, col2: number): number {
        let bottom_right = this.prefix_sum[row2][col2];
        let above = 0;
        let left = 0;
        let top_left = 0;
        if(row1 > 0)
            above = this.prefix_sum[row1- 1][col2];
        if(col1 > 0)
            left = this.prefix_sum[row2][col1 - 1];
        if(row1 > 0 && col1 > 0)
            top_left = this.prefix_sum[row1 - 1][col1 - 1];
        return bottom_right - above - left + top_left;

    }
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * var obj = new NumMatrix(matrix)
 * var param_1 = obj.sumRegion(row1,col1,row2,col2)
 */